File: api/Device.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/enums"
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// Device 对应 Device.py。
type Device struct {
	*session.ApiGroup
}

// NewDevice 创建 Device API 分组。
func NewDevice(s *session.Session) *Device {
	return &Device{ApiGroup: session.NewApiGroup(s)}
}

// Information 获取设备信息。对应 device/information。
func (d *Device) Information() (map[string]interface{}, error) {
	return d.S.Get("device/information", nil, "api")
}

// AutorunVersion 获取 autorun 版本。对应 device/autorun-version。
func (d *Device) AutorunVersion() (map[string]interface{}, error) {
	return d.S.Get("device/autorun-version", nil, "api")
}

// DeviceFeatureSwitch 获取设备功能开关状态。对应 device/device-feature-switch。
func (d *Device) DeviceFeatureSwitch() (map[string]interface{}, error) {
	return d.S.Get("device/device-feature-switch", nil, "api")
}

// BasicInformation 获取基本设备信息。对应 device/basic_information。
func (d *Device) BasicInformation() (map[string]interface{}, error) {
	return d.S.Get("device/basic_information", nil, "api")
}

// SetBasicInformation 设置基本设备信息。对应 device/basic_information (post_set)。
func (d *Device) SetBasicInformation(restoreDefaultStatus bool) (interface{}, error) {
	return d.S.PostSet("device/basic_information", map[string]interface{}{
		"restore_default_status": boolToInt(restoreDefaultStatus),
	}, false, "api", false, false)
}

// Basicinformation 获取基本设备信息（另一端点）。对应 device/basicinformation。
func (d *Device) Basicinformation() (map[string]interface{}, error) {
	return d.S.Get("device/basicinformation", nil, "api")
}

// UsbTetheringSwitch 获取 USB 网络共享开关状态。对应 device/usb-tethering-switch。
func (d *Device) UsbTetheringSwitch() (map[string]interface{}, error) {
	return d.S.Get("device/usb-tethering-switch", nil, "api")
}

// BootTime 获取设备启动时间。对应 device/boot_time。
func (d *Device) BootTime() (map[string]interface{}, error) {
	return d.S.Get("device/boot_time", nil, "api")
}

// SetControl 控制设备电源状态。对应 device/control (post_set)。
func (d *Device) SetControl(control enums.ControlMode) (interface{}, error) {
	return d.S.PostSet("device/control", map[string]interface{}{
		"Control": int(control),
	}, false, "api", false, false)
}

// Signal 获取设备信号信息。对应 device/signal。
func (d *Device) Signal() (map[string]interface{}, error) {
	return d.S.Get("device/signal", nil, "api")
}

// AntennaStatus 获取设备天线状态。对应 device/antenna_status。
func (d *Device) AntennaStatus() (map[string]interface{}, error) {
	return d.S.Get("device/antenna_status", nil, "api")
}

// GetAntennaSettings 获取设备天线设置。对应 device/antenna_settings。
func (d *Device) GetAntennaSettings() (map[string]interface{}, error) {
	return d.S.Get("device/antenna_settings", nil, "api")
}

// SetAntennaSettings 设置设备天线设置。对应 device/antenna_settings (post_set)。
func (d *Device) SetAntennaSettings(antennaType enums.AntennaType) (interface{}, error) {
	return d.S.PostSet("device/antenna_settings", map[string]interface{}{
		"antenna_type": int(antennaType),
	}, false, "api", false, false)
}

// AntennaType 获取设备天线类型。对应 device/antenna_type。
func (d *Device) AntennaType() (map[string]interface{}, error) {
	return d.S.Get("device/antenna_type", nil, "api")
}

// AntennaSetType 获取设备天线设置类型。对应 device/antenna_set_type。
func (d *Device) AntennaSetType() (map[string]interface{}, error) {
	return d.S.Get("device/antenna_set_type", nil, "api")
}

// Logsetting 获取设备日志设置。对应 device/logsetting。
func (d *Device) Logsetting() (map[string]interface{}, error) {
	return d.S.Get("device/logsetting", nil, "api")
}

// Logport 获取设备日志端口。对应 device/logport。
func (d *Device) Logport() (map[string]interface{}, error) {
	return d.S.Get("device/logport", nil, "api")
}

// Datalock 获取设备数据锁状态。对应 device/datalock。
func (d *Device) Datalock() (map[string]interface{}, error) {
	return d.S.Get("device/datalock", nil, "api")
}

// Vendorname 获取设备厂商名称。对应 device/vendorname (post_get)。
// 注意：部分不支持此端点的设备会破坏会话，谨慎使用。
func (d *Device) Vendorname(lang string) (map[string]interface{}, error) {
	return d.S.PostGet("device/vendorname", map[string]interface{}{
		"language": lang,
	}, false, "api", false, false)
}

// Mode 设置设备模式（可开启 telnet/调试模式/生产模式，见 enums.Mode）。
// 对应 device/mode (post_set)。
func (d *Device) Mode(mode enums.Mode) (interface{}, error) {
	return d.S.PostSet("device/mode", map[string]interface{}{
		"mode": int(mode),
	}, false, "api", false, false)
}

// CompressLogfile 返回归档日志文件链接。对应 device/compresslogfile。
func (d *Device) CompressLogfile() (map[string]interface{}, error) {
	return d.S.Get("device/compresslogfile", nil, "api")
}

// SecCellInfo 获取辅小区（载波聚合 SCell）信息。对应 device/seccellinfo。
// 5G CPE 专有端点，返回类似 "ARFCN,Band,BW(可选),PCI,RSRP,RSRQ,RSSI,SINR;..."
// 的 CSV 风格字段（nrseccell_list / lteseccell_list 等）。部分固件返回空响应。
func (d *Device) SecCellInfo() (map[string]interface{}, error) {
	return d.S.Get("device/seccellinfo", nil, "api")
}

// NbrCellInfo 获取邻小区（邻区/Neighbor Cell）信息。对应 device/nbrcellinfo。
// 5G CPE 专有端点，返回类似 "ARFCN,Band,PCI,RSRP,RSRQ,RSSI,SINR;..."
// 的 CSV 风格字段（nbrcell_nrlist / nbrcell_ltelist 等）。部分固件返回空响应。
func (d *Device) NbrCellInfo() (map[string]interface{}, error) {
	return d.S.Get("device/nbrcellinfo", nil, "api")
}

File: src/api/device.rs
//! Device API group (`api/Device.py`).

use serde_json::Value;

use crate::connection::Connection;
use crate::enums::device::{AntennaType, ControlMode, Mode};
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::XmlMap;

use super::{get_value, post_get_value, post_set};

/// Device API group.
pub struct Device<'a> {
    conn: &'a Connection,
}

impl<'a> Device<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        Device { conn }
    }

    /// `device/information`. Device information.
    pub fn information(&self) -> Result<Value> {
        get_value(self.conn, "api/device/information")
    }

    /// `device/autorun-version`. Autorun version.
    pub fn autorun_version(&self) -> Result<Value> {
        get_value(self.conn, "api/device/autorun-version")
    }

    /// `device/device-feature-switch`. Feature switch status.
    pub fn device_feature_switch(&self) -> Result<Value> {
        get_value(self.conn, "api/device/device-feature-switch")
    }

    /// `device/basic_information`. Basic device information.
    pub fn basic_information(&self) -> Result<Value> {
        get_value(self.conn, "api/device/basic_information")
    }

    /// Restore the basic-information factory defaults.
    pub fn set_basic_information(&self, restore_default_status: bool) -> Result<String> {
        let body: XmlMap = map_of([(
            "restore_default_status",
            if restore_default_status {
                "1".to_string()
            } else {
                "0".to_string()
            },
        )]);
        post_set(self.conn, "api/device/basic_information", &body)
    }

    /// `device/basicinformation`. Basic device information (alternate endpoint).
    pub fn basicinformation(&self) -> Result<Value> {
        get_value(self.conn, "api/device/basicinformation")
    }

    /// `device/usb-tethering-switch`. USB tethering switch status.
    pub fn usb_tethering_switch(&self) -> Result<Value> {
        get_value(self.conn, "api/device/usb-tethering-switch")
    }

    /// `device/boot_time`. Device boot time.
    pub fn boot_time(&self) -> Result<Value> {
        get_value(self.conn, "api/device/boot_time")
    }

    /// Control the device power state (reboot, reset, power off, ...).
    pub fn set_control(&self, control: ControlMode) -> Result<String> {
        let body: XmlMap = map_of([("Control", control.as_i32().to_string())]);
        post_set(self.conn, "api/device/control", &body)
    }

    /// `device/signal`. Signal information.
    pub fn signal(&self) -> Result<Value> {
        get_value(self.conn, "api/device/signal")
    }

    /// `device/antenna_status`. Antenna status.
    pub fn antenna_status(&self) -> Result<Value> {
        get_value(self.conn, "api/device/antenna_status")
    }

    /// `device/antenna_settings`. Antenna settings.
    pub fn get_antenna_settings(&self) -> Result<Value> {
        get_value(self.conn, "api/device/antenna_settings")
    }

    /// Configure the antenna type.
    pub fn set_antenna_settings(&self, antenna_type: AntennaType) -> Result<String> {
        let body: XmlMap = map_of([("antenna_type", antenna_type.as_i32().to_string())]);
        post_set(self.conn, "api/device/antenna_settings", &body)
    }

    /// `device/antenna_type`. Antenna type.
    pub fn antenna_type(&self) -> Result<Value> {
        get_value(self.conn, "api/device/antenna_type")
    }

    /// `device/antenna_set_type`. Antenna set type.
    pub fn antenna_set_type(&self) -> Result<Value> {
        get_value(self.conn, "api/device/antenna_set_type")
    }

    /// `device/logsetting` (reverse engineered, unknown purpose).
    pub fn logsetting(&self) -> Result<Value> {
        get_value(self.conn, "api/device/logsetting")
    }

    /// `device/logport`. Log port.
    pub fn logport(&self) -> Result<Value> {
        get_value(self.conn, "api/device/logport")
    }

    /// `device/datalock`. Data lock status.
    pub fn datalock(&self) -> Result<Value> {
        get_value(self.conn, "api/device/datalock")
    }

    /// `device/vendorname` (POST). Vendor name; may break unsupported devices.
    pub fn vendorname(&self, lang: &str) -> Result<Value> {
        let body: XmlMap = map_of([("language", lang.to_string())]);
        post_get_value(self.conn, "api/device/vendorname", &body)
    }

    /// Set the device mode (debug / telnet / production).
    pub fn mode(&self, mode: Mode) -> Result<String> {
        let body: XmlMap = map_of([("mode", mode.as_i32().to_string())]);
        post_set(self.conn, "api/device/mode", &body)
    }

    /// `device/compresslogfile`. Link to the archived log file.
    pub fn compress_logfile(&self) -> Result<Value> {
        get_value(self.conn, "api/device/compresslogfile")
    }

    /// `device/seccellinfo`. 5G secondary (carrier-aggregation) cell info.
    ///
    /// Supplementary endpoint from
    /// Brovi-Huawei-5G-CPE-Manager.
    pub fn get_sec_cell_info(&self) -> Result<Value> {
        get_value(self.conn, "api/device/seccellinfo")
    }

    /// `device/nbrcellinfo`. 5G neighbour-cell info.
    ///
    /// Supplementary endpoint from
    /// Brovi-Huawei-5G-CPE-Manager.
    pub fn get_nbr_cell_info(&self) -> Result<Value> {
        get_value(self.conn, "api/device/nbrcellinfo")
    }
}
